staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go

```diff
@@ -0,0 +1,295 @@
+		crdSpecsByGV: map[string]map[string]*customResourceDefinitionData{},
+	}
+
+	crdInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addCustomResourceDefinition,
+		UpdateFunc: c.updateCustomResourceDefinition,
+		DeleteFunc: c.deleteCustomResourceDefinition,
+	})
+
+	c.syncFn = c.sync
+	return c
+}
+
+// Run sets openAPIAggregationManager and starts workers
+func (c *Controller) Run(staticSpec *spec.Swagger, openAPIV3Service *handler3.OpenAPIService, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+	defer klog.Infof("Shutting down OpenAPI V3 controller")
+
+	klog.Infof("Starting OpenAPI V3 controller")
+
+	c.staticSpec = staticSpec
+	c.openAPIV3Service = openAPIV3Service
+
+	if !cache.WaitForCacheSync(stopCh, c.crdsSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	crds, err := c.crdLister.List(labels.Everything())
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to initially list all CRDs: %v", err))
+		return
+	}
+	for _, crd := range crds {
+		if !apiextensionshelpers.IsCRDConditionTrue(crd, apiextensionsv1.Established) {
+			continue
+		}
+		for _, v := range crd.Spec.Versions {
+			if !v.Served {
+				continue
+			}
+			c.buildV3Spec(crd, crd.Name, v.Name)
+		}
+	}
+
+	// only start one worker thread since its a slow moving API
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	// log slow aggregations
+	start := time.Now()
+	defer func() {
+		elapsed := time.Since(start)
+		if elapsed > time.Second {
+			klog.Warningf("slow openapi aggregation of %q: %s", key.(string), elapsed)
+		}
+	}()
+
+	err := c.syncFn(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with: %v", key, err))
+	c.queue.AddRateLimited(key)
+	return true
+}
+
+func (c *Controller) sync(name string) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	crd, err := c.crdLister.Get(name)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	if errors.IsNotFound(err) || !apiextensionshelpers.IsCRDConditionTrue(crd, apiextensionsv1.Established) {
+		c.deleteCRD(name)
+		return nil
+	}
+
+	for _, v := range crd.Spec.Versions {
+		if !v.Served {
+			continue
+		}
+		c.buildV3Spec(crd, name, v.Name)
+	}
+
+	return nil
+}
+
+func (c *Controller) deleteCRD(name string) {
+	for gvString, crdList := range c.crdSpecsByGV {
+		needOpenAPIUpdate := false
+		for crdName, crd := range crdList {
+			if crd.name == name {
+				needOpenAPIUpdate = true
+				delete(crdList, crdName)
+			}
+		}
+		if needOpenAPIUpdate {
+			if len(crdList) == 0 {
```

Comment:
this can never happen

Reply:
If there are no crds left in a group/version, the group/version should be deleted.

Reply:
The variable naming wasn't very clear. I've updated the name of this to crdListByGV so it's clear that this is only checking for the scenario of deleting the last CRD in a group rather than deleting all CRDs.